src/Microsoft.AspNet.OData.Shared/Query/Expressions/SelectExpandPathExtensions.cs

```diff
@@ -0,0 +1,128 @@
+            {
+                throw new ArgumentNullException("expandPath");
+            }
+
+            // In fact, ODataExpandPath constructor verifies the supporting segments, we add the verification here for double check.
+            return GetFirstNonTypeCastSegment(expandPath,
+                //The middle segment should be "TypeSegment" or "PropertySegment".
+                m => m is PropertySegment || m is TypeSegment,
+                // The last segment could be "NavigationPropertySegment"
+                s => s is NavigationPropertySegment,
+                out remainingSegments);
+        }
+
+        private static ODataPathSegment GetFirstNonTypeCastSegment(ODataPath path,
+            Func<ODataPathSegment, bool> middleSegmentPredict,
+            Func<ODataPathSegment, bool> lastSegmentPredict,
+            out IList<ODataPathSegment> remainingSegments) // could be null
+        {
+            Contract.Assert(path != null);
+
+            remainingSegments = null;
+            ODataPathSegment firstNonTypeSegment = null;
+            int lastIndex = path.Count() - 1;
+            int index = 0;
+            foreach (var segment in path)
+            {
+                if (index == lastIndex)
+                {
+                    // Last segment
+                    if (!lastSegmentPredict(segment))
+                    {
+                        throw new ODataException(Error.Format(SRResources.InvalidLastSegmentInSelectExpandPath, segment.GetType().Name));
+                    }
+                }
+                else
+                {
+                    // middle segment
+                    if (!middleSegmentPredict(segment))
+                    {
+                        throw new ODataException(Error.Format(SRResources.InvalidSegmentInSelectExpandPath, segment.GetType().Name));
+                    }
+                }
+
+                index++;
+
+                if (firstNonTypeSegment != null)
+                {
+                    if (remainingSegments == null)
+                    {
+                        remainingSegments = new List<ODataPathSegment>();
+                    }
+
+                    remainingSegments.Add(segment);
+                    continue;
+                }
+
+                // Theorically, a path like:  "~/NS.BaseType/NS.SubType1/NS.SubType2/PropertyOnSubType2" is valid(?) but not allowed.
+                // However, the functionality of above path is same as "~/NS.SubType2/PropertyOnSubType2".
+                // So, Let's only care about the last segment in the leading segments.
+                // if we have the leading segments, and the last segment must be the type segment and it's verified.
+                if (segment is TypeSegment)
+                {
+                    // do nothing here, just skip the leading type segment
+                }
+                else
+                {
+                    firstNonTypeSegment = segment;
+                }
```

Comment:
if (! (segment is TypeSegment)) instead of if else?

Reply:
It's by design. I leave the codes like these for better understanding/maintenance. 

---
In reply to: [352416339](https://github.com/OData/WebApi/pull/1941#discussion_r352416339) [](ancestors = 352416339)